File: solutions/go/custom-set/1/custom_set.go
package stringset

// Implement Set as a collection of unique string values.
//
// For Set.String, use '{' and '}', output elements as double-quoted strings
// safely escaped with Go syntax, and use a comma and a single space between
// elements. For example, a set with 2 elements, "a" and "b", should be formatted as {"a", "b"}.
// Format the empty set as {}.

// Define the Set type here.

type Set struct {
	elements *[]string
}

func New() Set {
	empty := make([]string, 0)
	return Set{&empty}
}

func NewFromSlice(l []string) Set {
	var newSet Set
	newSet = New()
	for _, elem := range l {
		newSet.Add(elem)
	}
	return newSet
}

func (s Set) String() string {
	var str string
	str = "{"
	for i, elem := range *s.elements {
		if i == len(*s.elements)-1 {
			str += "\"" + elem + "\""
		} else {
			str += "\"" + elem + "\", "
		}
	}
	str += "}"
	return str
}

func (s Set) IsEmpty() bool {
	if len(*s.elements) == 0 {
		return true
	}

	return false
}

func (s Set) Has(elem string) bool {
	for _, e := range *s.elements {
		if e == elem {
			return true
		}
	}

	return false
}

func (s Set) Add(elem string) {
	if !s.Has(elem) {
		*s.elements = append(*s.elements, elem)
	}
}

func Subset(s1, s2 Set) bool {
	var isSubset bool
	isSubset = true
	for _, e1 := range *s1.elements {
		if !s2.Has(e1) {
			isSubset = false
			return isSubset
		}
	}
	return isSubset
}

func Disjoint(s1, s2 Set) bool {
	for _, e1 := range *s1.elements {
		if s2.Has(e1) {
			return false
		}
	}
	return true
}

func Equal(s1, s2 Set) bool {
	if len(*s1.elements) != len(*s2.elements) {
		return false
	}

	for _, e1 := range *s1.elements {
		if !s2.Has(e1) {
			return false
		}
	}
	return true
}

func Intersection(s1, s2 Set) Set {
	var intersection Set
	intersection = New()
	for _, e1 := range *s1.elements {
		if s2.Has(e1) {
			intersection.Add(e1)
		}
	}
	return intersection
}

func Difference(s1, s2 Set) Set {
	var difference Set
	difference = New()
	for _, e1 := range *s1.elements {
		if !s2.Has(e1) {
			difference.Add(e1)
		}
	}
	return difference
}

func Union(s1, s2 Set) Set {
	var union Set
	union = New()
	for _, e1 := range *s1.elements {
		union.Add(e1)
	}
	for _, e2 := range *s2.elements {
		union.Add(e2)
	}
	return union
}

File: solutions/typescript/custom-set/1/custom-set.ts
export class CustomSet {
  setOfNumbers: number[] = []

  constructor(initial?: number[]) {
    this.setOfNumbers = initial || []
  }

  empty(): boolean{
    if (this.setOfNumbers.length === 0) {
      return true
    }
    return false
  }

  contains(element: number): boolean {
    if (this.setOfNumbers.includes(element)) {
      return true
    }

    return false
  }

  add(element: number): CustomSet {
    if (!this.setOfNumbers.includes(element)) {
      this.setOfNumbers.push(element)
    }
    return this
  }

  subset(other: CustomSet): boolean {
    if (this.setOfNumbers.length === 0) {
      return true
    }

    for (let i = 0; i < this.setOfNumbers.length; i++) {
      if (!other.setOfNumbers.includes(this.setOfNumbers[i])) {
        return false
      }
    }
    return true
  }

  disjoint(other: CustomSet): boolean {
    if (this.setOfNumbers.length === 0 && other.setOfNumbers.length === 0) {
      return true
    }
    if (this.setOfNumbers.length === 0 || other.setOfNumbers.length === 0) {
      return true
    }

    for (let i = 0; i < this.setOfNumbers.length; i++) {
      if (other.setOfNumbers.includes(this.setOfNumbers[i])) {
        return false
      }
    }
    return true
  }

  eql(other: CustomSet): boolean {
   if (this.setOfNumbers.length !== other.setOfNumbers.length) {
     return false
   }
   if (this.setOfNumbers.length === 0 && other.setOfNumbers.length === 0) {
     return true
   }

    for (let i = 0; i < this.setOfNumbers.length; i++) {
      if (!other.setOfNumbers.includes(this.setOfNumbers[i])) {
        return false
      }
    }
    return true
  }

  union(other: CustomSet): CustomSet {
    let unionSet = new CustomSet()
    for (let i = 0; i < this.setOfNumbers.length; i++) {
      unionSet.add(this.setOfNumbers[i])
    }
    for (let i = 0; i < other.setOfNumbers.length; i++) {
      unionSet.add(other.setOfNumbers[i])
    }
    return unionSet
  }

  intersection(other: CustomSet): CustomSet {
    let intersectionSet = new CustomSet()
    for (let i = 0; i < this.setOfNumbers.length; i++) {
      if (other.setOfNumbers.includes(this.setOfNumbers[i])) {
        intersectionSet.add(this.setOfNumbers[i])
      }
    }
    return intersectionSet
  }

  difference(other: CustomSet): CustomSet {
    let differenceSet = new CustomSet()
    for (let i = 0; i < this.setOfNumbers.length; i++) {
      if (!other.setOfNumbers.includes(this.setOfNumbers[i])) {
        differenceSet.add(this.setOfNumbers[i])
      }
    }

    return differenceSet
  }
}
